### PR TITLE
Make bind() optional and implement a generator for device discovery

### PR DESCRIPTION
**Pull request by felipediel**
## The problem
Many users are not being able to discover their devices in Home Assistant. This is happening because our local IP address lookup logic does not take into account that there may be multiple network interfaces for the same host.

If we bind() to the wrong interface we cannot communicate to the device. We already talked about this [here](https://github.com/mjg59/python-broadlink/pull/378). There I exposed the problem and why we should not rely on `socket.gethostbyname(socket.gethostname())` as a default option, but I didn't solve it completely. Now that discovering the device is mandatory for its configuration in Home Assistant, I think I better stop postponing and fix this for good.

I opted for the minimalist solution: is bind() really necessary? According to the Broadlink protocol, we should attach the source ip and port to the packet. But in fact, this is not really necessary. We did a lot of tests and we can communicate normally without this, which exists for debugging purposes I guess. So we are gonna avoid bind() unless the user ask us to do so by offering a local_ip_address argument. I will leave this option intact for compatibility and also as a safeguard in case we need this in the future.

I will take the opportunity to make discovery much faster by implementing a generator. So we don't need to wait for the timeout to be reached in order to access the results. This makes our configuration, initialization and updates much faster in Home Assistant.

## Breaking changes
- Change discover() behavior: always return a list, even if no timeout is given. We should use hello() for direct discovery. This abstraction helps with typing and automated tests, but the main reason I did this is because discover() with no timeout is bad, it blocks the execution if the address is wrong or the device is offline.

## Proposed changes
1. Remove local IP address lookup logic. Do not bind() if the user doesn't ask us to do so. This fixes:
- https://github.com/home-assistant/core/issues/40191#issuecomment-695803468
- https://github.com/home-assistant/core/issues/40267

2. Implement a generator to discover devices (we no longer need to wait for the timeout in order to get a list, we yield devices instantly).
```python3
import broadlink as blk
for d in blk.xdiscover():
    print(d)
```

3. Implement a hello function at `__init__.py`. This function sends a hello message to a host and generates a device with the response. This is like a xdiscover() with no multicast. It works even if the device is locked. Useful for generating devices with known hosts. This abstraction also helps a lot with setup and automated tests in Home Assistant.
```python3
import broadlink as blk
d = blk.hello("192.168.0.17")  # Device IP address.
```
4. Implement a hello method for the device class. This is useful for pings, name/lock status updates and checking if static information changed (e.g. if the MAC address or the type changes, the device information is corrupted).
```python3
import broadlink as blk

# This generates a device.
d = blk.hello("192.168.0.17")  # Device IP address.

# This updates the device.
d.hello()  # Send hello again and update name and lock status with the response.
```
If the device type or MAC address changed we raise a DeviceInformationError() to inform the client that this is not the device this object was supposed to control.

## Tests
- works for hosts with multiple interfaces: https://github.com/home-assistant/core/issues/40191#issuecomment-696079386
- works for VPNs: https://github.com/mjg59/python-broadlink/pull/378#issuecomment-696092120
- works fine in all the conditions I could reproduce (Docker, venv, multiple network interfaces).

### CODE REVIEW

File: broadlink/device.py [hunk 1]
```diff
@@ -2,13 +2,93 @@
...
+    discover_ip_address: str = '255.255.255.255',
+    discover_ip_port: int = 80,
+) -> Generator[HelloResponse, None, None]:
+    """Broadcast a hello message and yield responses."""
+    cs = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    cs.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+    cs.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
+
+    if local_ip_address:
+        cs.bind((local_ip_address, 0))
```
Comment (felipediel) [inline broadlink/device.py:28]: This line is very aggressive. It is also unnecessary, so I'm gonna make it optional.

File: broadlink/device.py [hunk 2]
```diff
@@ -2,13 +2,93 @@
...
+    packet[0x10] = int(subyear)
+    packet[0x11] = datetime.now().isoweekday()
+    packet[0x12] = datetime.now().day
+    packet[0x13] = datetime.now().month
+    packet[0x18] = int(address[3])
+    packet[0x19] = int(address[2])
+    packet[0x1a] = int(address[1])
+    packet[0x1b] = int(address[0])
+    packet[0x1c] = port & 0xff
+    packet[0x1d] = port >> 8
```
Comment (felipediel) [inline broadlink/device.py:67]: These bytes are not a big deal, they are only used for debugging.

File: broadlink/__init__.py [hunk 3]
```diff
@@ -118,91 +114,51 @@ def gendevice(
...
-        response = cs.recvfrom(1024)
-        responsepacket = bytearray(response[0])
-        host = response[1]
-        devtype = responsepacket[0x34] | responsepacket[0x35] << 8
-        mac = responsepacket[0x3f:0x39:-1]
-        name = responsepacket[0x40:].split(b'\x00')[0].decode('utf-8')
-        is_locked = bool(responsepacket[-1])
-        device = gendevice(devtype, host, mac, name=name, is_locked=is_locked)
-        cs.close()
-        return device
```
Comment (felipediel) [inline broadlink/__init__.py:187]: I'm gonna remove this to improve typing. This function will always return a list. If we want direct discovery we can use hello().

File: broadlink/exceptions.py [hunk 4]
```diff
@@ -88,10 +92,6 @@ class LengthError(SDKException):
     """Received data packet length error."""
 
 
-class DNSLookupError(SDKException):
-    """Failed to obtain local IP address."""
```
Comment (felipediel) [inline broadlink/exceptions.py:92]: We don't need this anymore. I feel stupid.

**Review by Danielhiversen — APPROVED** [review]
LGTM 🎉

**Pull request merged**